Assets/Scripts/Noise/PerlinNoise.cs

```diff
@@ -0,0 +1,210 @@
+    /// Bounds the perlin noise to a grid space and repeats that grid square.
+    /// </summary>
+    private int repeat = 0;
+
+    /// <summary>
+    /// Makes a perlin noise generator with given settings.
+    /// </summary>
+    /// <param name="repeat">Bounds noise to a repeating square of size repeat by repeat. If this 
+    /// value is set to zero, there is no repeat.</param>
+    /// /// <param name="seed">Seed value for random permutation. if set to zero a arbitrary seed will be used.</param>
+    public PerlinNoise(int repeat, int seed) {
+        int[] permutation = new int[256];
+
+        for (int i = 0; i < 256; i++) {
+            permutation[i] = i;
+        }
+        
+        System.Random random = seed == 0 ? new System.Random() : new System.Random(seed);
+        int n = permutation.Length;
+        for (int i = n - 1; i > 1; i--) {
+            int rnd = random.Next(i+1);
+            int value = permutation[rnd];
+            permutation[rnd] = permutation[i];
+            permutation[i] = value;
+        }
+
+        p = new int[512];
+        for(int x=0;x<512;x++) {
+            p[x] = permutation[x%256];
+        }
+
+        this.repeat = repeat;
+    }
+
+    /// <summary>
+    /// Calculates the Perlin Noise value at a given x, y, z.
+    /// </summary>
+    /// <param name="x">X coordinate</param>
+    /// <param name="y">Y coordinate</param>
+    /// <param name="z">Z coordinate</param>
+    /// <returns>Perlin noise value at the given coordinate.</returns>
+    public float Perlin(float x, float y, float z) {
+        // If we have any repeat on, change the coordinates to their "local" repetitions
+        if (repeat > 0) {
+            x = x % repeat;
+            y = y % repeat;
+            z = z % repeat;
+        }
+
+        // Calculate the "unit cube" that the point asked will be located in
+        // The left bound is ( |_x_|,|_y_|,|_z_| ) and the right bound is that
+        // plus 1.  Next we calculate the location (from 0.0 to 1.0) in that cube.
+        int xi = (int)x & 255;
+        int yi = (int)y & 255;
+        int zi = (int)z & 255;
+        float xf = x - (int)x;
+        float yf = y - (int)y;
+        float zf = z - (int)z;
+
+        // Compute faded values for smoothing
+        float u = Fade(xf);
+        float v = Fade(yf);
+        float w = Fade(zf);
+
+        // Perlin Noise hash function is used to get a unique value for every coordinate input.
+        // This is the hash for each of the 8 points on the unit cube around this point
+        int aaa, aba, aab, abb, baa, bba, bab, bbb;
+        aaa = p[p[p[    xi ]+    yi ]+    zi ];
+        aba = p[p[p[    xi ]+Inc(yi)]+    zi ];
+        aab = p[p[p[    xi ]+    yi ]+Inc(zi)];
+        abb = p[p[p[    xi ]+Inc(yi)]+Inc(zi)];
+        baa = p[p[p[Inc(xi)]+    yi ]+    zi ];
+        bba = p[p[p[Inc(xi)]+Inc(yi)]+    zi ];
+        bab = p[p[p[Inc(xi)]+    yi ]+Inc(zi)];
+        bbb = p[p[p[Inc(xi)]+Inc(yi)]+Inc(zi)];
```

Comment:
Room for a helper function? Maybe pull this into a helper function that returns the tuple of all the values. This function is rather large right now an abstracting some of this might be helpful.

Reply:
Yes, I moved this to a helper function.